source/common/http/filter/cors_filter.cc

```diff
@@ -0,0 +1,152 @@
+
+#include "envoy/http/codes.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+
+namespace Envoy {
+namespace Http {
+
+CorsFilter::CorsFilter(CorsFilterConfigConstSharedPtr config)
+    : config_(config), is_cors_request_(false) {}
+
+CorsFilter::~CorsFilter() {}
+
+FilterHeadersStatus CorsFilter::decodeHeaders(HeaderMap& headers, bool) {
+  origin_ = headers.Origin();
+  if (origin_ == nullptr || origin_->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  initialize();
+
+  if (!enabled()) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (!isOriginAllowed(origin_->value())) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  is_cors_request_ = true;
+
+  auto method = headers.Method();
+  if (method == nullptr || method->value().c_str() != Http::Headers::get().MethodValues.Options) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  auto requestMethod = headers.AccessControlRequestMethod();
+  if (requestMethod == nullptr || requestMethod->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  HeaderMapPtr response_headers{
+      new HeaderMapImpl{{Headers::get().Status, std::to_string(enumToInt(Http::Code::OK))}}};
+
+  response_headers->insertAccessControlAllowOrigin().value(*origin_);
+
+  if (allowCredentials()) {
+    response_headers->insertAccessControlAllowCredentials().value(
+        Http::Headers::get().CORSValues.True);
+  }
+
+  if (allowMethods() != "") {
+    response_headers->insertAccessControlAllowMethods().value(allowMethods());
+  }
+
+  if (allowHeaders() != "") {
+    response_headers->insertAccessControlAllowHeaders().value(allowHeaders());
+  }
+
+  if (exposeHeaders() != "") {
+    response_headers->insertAccessControlExposeHeaders().value(exposeHeaders());
+  }
+
+  if (maxAge() != "") {
+    response_headers->insertAccessControlMaxAge().value(maxAge());
+  }
+
+  decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterHeadersStatus CorsFilter::encodeHeaders(HeaderMap& headers, bool) {
+  if (!is_cors_request_) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  headers.insertAccessControlAllowOrigin().value(*origin_);
+  if (allowCredentials()) {
+    headers.insertAccessControlAllowCredentials().value(Http::Headers::get().CORSValues.True);
+  }
+
+  return FilterHeadersStatus::Continue;
+}
+
+void CorsFilter::setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+};
+
+void CorsFilter::initialize() {
+  route_cors_policy_ = decoder_callbacks_->route()->routeEntry()->corsPolicy();
+  virtual_host_cors_policy_ = decoder_callbacks_->route()->routeEntry()->virtualHost().corsPolicy();
+}
+
+bool CorsFilter::isOriginAllowed(Http::HeaderString& origin) {
+  for (const auto& o : allowOrigin()) {
+    if (o == "*" || origin == o.c_str()) {
```

Comment:
I believe this is the correct solution for envoy but it reminds me to try to sell @mattklein123 on the wonders of hash maps for things like this and header storage :-)

Reply:
I was thinking of using hash maps for this, but sticked to what I found in the code already. I imagined somebody maybe sticking 10 domains in there might be much already. Could be wrong though. I could change it in a future pr. What's the back story of selling Matt to use them? Does he dislike hashmaps? :)

Reply:
I'm sorry I don't have any context. I have nothing against maps (we use them in many places). Will need more info to offer guidance.

Reply:
Looking at this review now. If you want to use a map here, please use a map. I'm not sure where anyone is getting the impression that I don't like maps! :)